### ISSUE TITLE
parakeet-tdt-0.6b-v3-coreml: window-composition instability corrupts words (right-context sensitivity) in the fixed 15s non-causal export

### ISSUE THREAD

**Issue by holovchenko** [FluidInference/FluidAudio#760]
### Summary
On windowed/long-form transcription, `parakeet-tdt-0.6b-v3-coreml` intermittently drops or mangles a word when that word lands within roughly the first ~5.5s of the 15s encoder window covering it. The identical weights via NeMo (PyTorch) and parakeet-mlx transcribe the same audio correctly, so this is specific to the CoreML export + windowing path, not the model weights.

### Reproduction (deterministic; Ukrainian; `.cpuOnly`; `ASRConfig(melChunkContext: false)`)
118s clip. Speaker says "…вимкнути фоновий процес…" at 43.6–46.6s.
- FluidAudio (whole clip, chunked): "…вимкнути там кнопкою **фоновиц**" — "процес" dropped, "фоновий"→"фоновиц".
- NeMo `nvidia/parakeet-tdt-0.6b-v3` (whole audio): "…**фоновий процес**" ✓
- parakeet-mlx `mlx-community/parakeet-tdt-0.6b-v3` (whole audio): "…**фоновий процес**" ✓

### Isolated cause — a left-context cliff
Single-shot inference (no chunk merge), varying the left audio context before the phrase:

| left context before phrase | output |
|---|---|
| 1s / 3s / 4s | mangled / "фоновиц" (word dropped) |
| 5.5s / 7s | "фоновий процес" ✓ |

The CoreML Encoder is stateless and fixed-shape (`mel [1,128,1501]` = 15.01s, `hasShapeFlexibility: 0`, empty `stateSchema`); the 15s window is baked at trace time (`mobius .../coreml/convert-parakeet.py`, `max_audio_seconds = 15.0`). With `melChunkContext: false`, the only left context a window gets is the sliding overlap (default 2s), so a word near a window's start sits below the ~5.5s the encoder needs and is dropped/mangled.

### Widening the sliding overlap does NOT fix it
Raising `ChunkProcessor.overlapSeconds` 2→6s (the covering window then has ~7.76s of left context) still drops the word: the merge/token-selection picks the adjacent *short-context* window's tokens in the overlap zone. It also costs ~8–24% compute on multi-window clips.

### Suggested directions
1. Export the encoder with a larger context window (e.g. 30s) or a stateful/streaming variant, so positions aren't starved of left context.
2. Make merge/token-selection in overlap zones prefer the window where a token had more left context (i.e. is deeper into its window), rather than the window where it sits near the start.

### Environment
FluidAudio 0.15.3+, `parakeet-tdt-0.6b-v3-coreml`, macOS 26 / iOS, `.cpuOnly`, `melChunkContext: false`.

---

### Correction / refined mechanism (further ablation)

Additional isolation shows this is **window-composition instability**, not purely a *left*-context cliff. Holding the word fixed and truncating the surrounding audio:

- word decodes **correctly** with adequate left context AND ≤~2s of trailing audio after it in the window;
- word **corrupts** ("фоновиц", "процес" dropped) once there is ≥~4s of audio *after* the word inside the same 15s window.

So the non-causal (bidirectional-within-window) encoder's encoding of a word is unstable to what *follows* it in the fixed window, not just what precedes it. Widening the sliding overlap therefore can't help (the word sits in a single window with no competing decode), and simply enlarging the window may add more trailing context rather than less. Reference runtimes with full-file attention (NeMo, parakeet-mlx) don't exhibit this. This points to the fixed-window / non-causal CoreML export (possibly interacting with INT8 quantization) rather than the chunk-merge or the weights.

**Comment by Alex-Wengg** [comment]
@holovchenko does this issue still persist

**Comment by Alex-Wengg** [comment]
@holovchenko if so please provide the audio transcription

**Comment by hiSandog** [comment]
A regression fixture could hold the target word and its left context constant while varying only the appended right-context audio and silence length. Reporting token timestamps or logits around the corrupted boundary would help distinguish window-stitching errors from genuine non-causal decoder sensitivity.

**Comment by holovchenko** [comment]
@Alex-Wengg Yes — still reproduces on `main` @ 5390df97 (v0.15.5 + 32 commits), stock `parakeet-tdt-0.6b-v3-coreml` (int8, 15 s window), CLI `transcribe`.

Audio attached below. All three cuts are **≤ 15 s**, so each decodes in a **single window** — no chunk merge, no seam, no `mergeChunks` involvement. Same source recording (16 kHz mono, Ukrainian), same word held fixed; the only thing that varies is how much audio sits **after** the word inside the window.

## Ablation

| file | left context before the word | audio after the word | decoded |
|---|---|---|---|
| `repro760-A-left11.3s-tail2.0s.wav` | 11.3 s | 2.0 s | `…кнопкою **фоновий процес**. Але додати.` ✅ |
| `repro760-B-left7.3s-tail6.0s.wav` | 7.3 s | 6.0 s | `…кнопкою **фоновиц**. Але додаток при цьому лишається у фоні. І коли я` ❌ |
| `repro760-C-control-left7.3s-tail2.0s.wav` | **7.3 s** | 2.0 s | `…кнопкою **фоновий процес**. Але додати.` ✅ |

C is the control: it has the **same left context as B** (7.3 s) and decodes the word correctly. So the failure is not a left-context cliff — it is driven by the ~4 extra seconds of audio *following* the word within the same fixed window. Expected text is `фоновий процес` ("background process"); B collapses it to a non-word `фоновиц` and deletes `процес`.

## Commands

```bash
fluidaudio transcribe repro760-A-left11.3s-tail2.0s.wav --language uk
fluidaudio transcribe repro760-B-left7.3s-tail6.0s.wav  --language uk
fluidaudio transcribe repro760-C-control-left7.3s-tail2.0s.wav --language uk
```

Invariant to the knobs I could vary:
- identical with and without `--no-mel-context`
- identical with `--language uk` and with no language hint (without the hint B also renders the brand name in Latin script, but the corrupted word is byte-identical)
- deterministic across repeated runs on the same machine (M-series, macOS 15)

## In the full recording

The same word corrupts in the original 118 s file at **t ≈ 45.28–46.96 s**, with `confidence 0.885` — high enough that no confidence filter would flag it, which is what makes this class expensive for us: the output is silently wrong, not visibly uncertain.

## What I did not test

Whether this is specific to the int8 encoder quantization (I only ran the int8 export), and whether the v2 export behaves the same. Reference runtimes with full-file attention (NeMo, parakeet-mlx) do not show it, which is what pointed me at the fixed-window non-causal CoreML export rather than the weights.

## Audio

https://github.com/holovchenko/FluidAudio/releases/tag/repro-760-audio — three WAVs, direct download links:

- https://github.com/holovchenko/FluidAudio/releases/download/repro-760-audio/repro760-A-left11.3s-tail2.0s.wav
- https://github.com/holovchenko/FluidAudio/releases/download/repro-760-audio/repro760-B-left7.3s-tail6.0s.wav
- https://github.com/holovchenko/FluidAudio/releases/download/repro-760-audio/repro760-C-control-left7.3s-tail2.0s.wav
